File: old_session/session_1/_20/_20_valid_parentheses_redo.java
/*
20. Valid Parentheses

    Total Accepted: 138038
    Total Submissions: 443232
    Difficulty: Easy

Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
determine if the input string is valid.

The brackets must close in the correct order,
"()" and "()[]{}" are all valid but "(]" and "([)]" are not.
*/
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        Map<Character, Character> par = new HashMap<Character, Character>();
        par.put('(', ')');
        par.put('[', ']');
        par.put('{', '}');

        for (char c: s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{')
                stack.push(c);
            else if (!stack.empty() && par.get(stack.peek()) == c)
                stack.pop();
            else
                return false;
        }
        return stack.empty();
    }
}

File: old_session/session_1/_20/_20_valid_parentheses_redo.py
"""
20. Valid Parentheses

Given a string containing just the characters '(', ')', '{', '}',
'[' and ']', determine if the input string is valid.

The brackets must close in the correct order,
"()" and "()[]{}" are all valid but "(]" and "([)]" are not.
"""
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack, i, left, right = [], 0, '([{', ')]}'

        for c in s:
            if c in left: stack += c,
            elif stack and right.index(c) == left.index(stack[-1]): stack.pop()
            else: return False

        return len(stack) == 0
